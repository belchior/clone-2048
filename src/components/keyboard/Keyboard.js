import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { parseShortcut } from './parseShortcut';


let userShortcuts;
const setUserShortcuts = shortcuts => {
  userShortcuts = shortcuts.reduce((list, item) => {
    const parsedShortcut = parseShortcut(item.shortcut);
    return parsedShortcut.keys.length > 0 ? list.concat({ ...item, ...parsedShortcut }) : list;
  }, []);
};

const matchWith = keyboardEvent => userKey => (
  userKey.altKey === keyboardEvent.altKey &&
  userKey.ctrlKey === keyboardEvent.ctrlKey &&
  userKey.metaKey === keyboardEvent.metaKey &&
  userKey.shiftKey === keyboardEvent.shiftKey &&
  userKey.keys.indexOf(keyboardEvent.key) >= 0 &&
  userKey.keys.indexOf(keyboardEvent.location) >= 0
);

const keyupHandler = keyboardEvent => {
  userShortcuts
    .filter(matchWith(keyboardEvent))
    .map(shortcut => shortcut.action());
};

export class Keyboard extends Component {
  componentDidMount() {
    const { targetSelector } = this.props;
    if (targetSelector) this.setDOMEvent(keyupHandler, targetSelector);
  }

  componentDidUpdate() {
    const { targetSelector } = this.props;
    if (targetSelector) this.setDOMEvent(keyupHandler, targetSelector);
  }

  render() {
    const { shortcuts, targetSelector, children } = this.props;
    setUserShortcuts(shortcuts);
    if (!targetSelector) return this.setSyntheticEvent(children, keyupHandler);
    return children;
  }

  setDOMEvent(handler, selector) {
    const target = document.querySelector(selector);

    // the only way to add keyboard events to a target diferent from html or body
    // is adding a tabIndex attribute to it.
    if (
      ['HTML', 'BODY'].indexOf(target.nodeName) >= 0 &&
      !target.getAttribute('tabIndex')
    ) target.setAttribute('tabIndex', -1);


    // There is no way to check if a event was attached before
    target.removeEventListener('keyup', handler);
    target.addEventListener('keyup', handler);
  }

  setSyntheticEvent(children, handler) {
    if (!children) return;
    const childList = Array.isArray(children) ? children : [children];

    return React.Children.map(childList, child => {
      const props = { ...child.props, onKeyUp: handler, tabIndex: -1};
      return React.cloneElement(child, props, props.children);
    });
  }
}

Keyboard.propTypes = {
  shortcuts: PropTypes.array.isRequired,
  children: PropTypes.any,
  targetSelector: PropTypes.string,
};
