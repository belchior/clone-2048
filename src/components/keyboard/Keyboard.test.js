import React, { Component } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { Keyboard } from './Keyboard';

const renderer = new ShallowRenderer();


test('Keyboard should be instantiable', () => {
  const keyboard = new Keyboard();

  expect(keyboard instanceof Keyboard).toBe(true);
});

test('Keyboard should be extended from React Component', () => {
  const keyboard = new Keyboard();

  expect(keyboard instanceof Component).toBe(true);
});

test('Keyboard should render without crashing', () => {
  const props = {
    shortcuts: [
      {shortcut: 'ctrl+d', action: () => {}}
    ]
  };
  const tree = renderer.render(
    <Keyboard {...props}><div className="test">test</div></Keyboard>
  );

  // console.error(tree);
  expect(tree).toMatchSnapshot();
});
