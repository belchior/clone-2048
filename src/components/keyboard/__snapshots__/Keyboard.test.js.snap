// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard should render without crashing 1`] = `
Array [
  <div
    className="test"
    onKeyUp={[Function]}
    tabIndex={-1}
>
    test
</div>,
]
`;
